doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1747,69 +1697,74 @@ name of the variable (here :g:`n`) is chosen based on :g:`T`.
-   same existential variable are counted multiple times.  Because this
-   variant is not robust to slight changes in the goal, its use is
-   strongly discouraged.
+   same existential variable are counted multiple times.  Using this form
+   is discouraged because slight changes to the goal may change the needed index,
+   causing a maintenance issue.
 
-.. tacv:: instantiate ( @natural := @term ) in @ident
-          instantiate ( @natural := @term ) in ( value of @ident )
-          instantiate ( @natural := @term ) in ( type of @ident )
+   Advanced users may want to define and use an Ltac tactic to get more consistent
+   behavior, such as:
 
-   These allow to refer respectively to existential variables occurring in a
-   hypothesis or in the body or the type of a local definition (named :n:`@ident`).
-
-.. tacv:: instantiate
-
-   This tactic behaves functionally as :tacn:`idtac`.
+   .. coqdoc::
 
-   .. deprecated:: 8.16
+      Ltac instantiate_ltac_variable ev term :=
+        let H := fresh in
+        pose ev as H;
+        instantiate (1 := term) in (value of H);
+        clear H.
 
-.. tacn:: admit
-   :name: admit
+   :n:`in @ident`
+     Selects the hypothesis :n:`@ident`.
 
-   This tactic allows temporarily skipping a subgoal so as to
-   progress further in the rest of the proof. A proof containing admitted
-   goals cannot be closed with :cmd:`Qed` but only with :cmd:`Admitted`.
+   :n:`in %|- *`
+     Selects the goal.  This is the default behavior.
 
-.. tacv:: give_up
+   :n:`in ( type of @ident )`
+     Selects existential variables in the type of the
+     :term:`local definition <context-local definition>` :n:`@ident`.
+     (The body is not included.)
 
-   Synonym of :tacn:`admit`.
+   :n:`in ( value of @ident )`
+     Selects existential variables in the body of the
+     :term:`local definition <context-local definition>` :n:`@ident`.
+     (The type is not included.)
 
-.. tacn:: absurd @term
-   :name: absurd
+.. tacn:: absurd @one_type
 
-   This tactic applies to any goal. The argument term is any proposition
+   :n:`@one_type` is any proposition
    :g:`P` of type :g:`Prop`. This tactic applies False elimination, that is it
    deduces the current goal from False, and generates as subgoals :g:`∼P` and
    :g:`P`. It is very useful in proofs by cases, where some cases are
    impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
    local context.
 
-.. tacn:: contradiction
-   :name: contradiction
+.. tacn:: contradiction {? @one_term_with_bindings }
 
-   This tactic applies to any goal. The contradiction tactic attempts to
-   find in the current context (after all intros) a hypothesis that is
-   equivalent to an empty inductive type (e.g. :g:`False`), to the negation of
-   a singleton inductive type (e.g. :g:`True` or :g:`x=x`), or two contradictory
-   hypotheses :g:`P` and :g:`not P`.
+   Tries to prove the current goal by finding a contradiction, such as
```

Comment:
Odd that the last line fails.  I would have expected it would introduce the negation of the statement.  Or is that unsound in  COC.  Also not sure whether the no-op is a feature or a bug.

```
Goal ~0=0 -> True.
intros.
contradiction H.  (* 0 = 0 *)
contradiction H.  (* no op *)
Abort.
Goal 0 = 0 -> True.
intros.
contradiction H. (* Not a contradiction. *)
```

Reply:
The no-op is normal. What `contradiction H` does in this case is that it replaces the current goal, whatever it is, with a goal asking you to prove the non-negated version of `H`. This is what happens in your example. But such a behavior is naturally "idempotent" (if you apply it a second time, you get the same result).

The behavior you expected for the last line is very reasonable, and I would have expected it as well, but it turns out that it is not the case (as I described in my second bullet point in https://github.com/coq/coq/pull/16498#discussion_r993126911). IMHO, it would be reasonable to extend `contradiction H` to not fail in this case. OTOH, there is already the `contradict` tactic which seems strictly more powerful for such uses.

Reply:
I'm reviewing uses of `contradiction` in the standard library. Apparently, it doesn't just look at the local context. This example shows that it can look into goal premises as well:

```coq
Lemma absurd_eq_bool : forall b b':bool, False -> b = b'.
Proof.
  contradiction.
Qed.
```

This is further confirmed by this other example:

```coq
Goal forall A B:Prop, A -> ~A -> B.
contradiction.
Qed.
```

Something else I just noticed is that `contradiction H` is less powerful than `contradiction`. Taking your example

```coq
Goal ~0=0 -> True.
intros.
contradiction H.
(* leaves the goal 0 = 0 which now requires an explicit call to reflexivity *)
Undo.
contradiction.
Qed.
```

Reply:
The only interesting case of using `contradiction` with an argument in the standard library is when using it with a term as argument (in the case it is used with a hypothesis, it can be replaced with `contradict`):

```coq
Require Import Arith.

Goal forall (A : Prop), 0 < 0 -> A.
intros. (* In this case, the call to intros is necessary *)
contradiction (Nat.lt_irrefl 0). (* Nat.lt_irrefl 0 is a proof of ~(0 < 0) *)
(* No more goals *)
Undo. (* Obviously this is strictly the same as doing: *)
pose proof (Nat.lt_irrefl 0); contradiction. (* No more goals *)
Restart. (* And other solutions are possible *)
intros * ?%Nat.lt_irrefl; contradiction.
Qed.
```

Reply:
> This example shows that [`contradiction` (no arg)] can look into goal premises as well.

The no argument form (only) does an `intros` first (line 142).  (I wonder a little how well-thought through the behavior is.)

The code is in `contradiction.ml`.  Start at the end of the file.  But hard for me to learn much from the code, maybe it makes more sense to you.  FWIW, we really should have some quality internals documentation for how the tactics code works--it seems pretty impenetrable even though what it does is, in many cases, probably simple.  Very few comments.

Reply:
> However, it is able to detect when it is solved by `assumption`.

In which example?  It didn't work in the example following your comment.

BTW, this commenting interface is awkward for commenting on longish comments with multiple points.  I've added some "JEF:" text to bottom of your comments.

Reply:
> So `contradiction H` is not able to detect that the remaining goal is solved by `reflexivity`.

Maybe the no-argument form is applying a reduction to address this?  Perhaps you can pick it out in the source code and tell me what reduction it is?

Reply:
> In which example? It didn't work in the example following your comment.

The example below the quoted sentence calls `contradiction H1` and `contradiction H2` and the tactic detects that the second hypothesis is need is in the context as well.